src/controllers/product.controller.ts

```diff
@@ -137,14 +152,66 @@ export class ProductController {
     @param.path.string('id') id: string,
     @requestBody() product: Product,
   ): Promise<void> {
-    await this.productRepository.replaceById(id, product);
+    if (product.previewImage?.startsWith('data:image')) {
+      await this.handleRawImageProductReplacement(product)
+    } else {
+      await this.productRepository.replaceById(id, product);
+    }
   }
 
   @del('/products/{id}')
   @response(204, {
     description: 'Product DELETE success',
   })
   async deleteById(@param.path.string('id') id: string): Promise<void> {
+    const product = await this.productRepository.findById(id);
+    if (product?.previewImage) {
+      this.deleteProductPreviewImage(product);
+    }
     await this.productRepository.deleteById(id);
   }
+
+  @get('/product-images/{filename}')
+  @oas.response.file()
+  sendFile(
+    @param.path.string('filename') fileName: string,
+    @inject(RestBindings.Http.RESPONSE) res: Response,
+  ) {
+    const filePath = path.resolve(this.imageStorageDirectory, fileName);
+    res.sendFile(filePath);
+    return res;
+  }
+
+  private async handleRawImageProductReplacement(product: Product) {
+    if (!product.previewImage) {
+      await this.productRepository.replaceById(product.id, product);
+      return;
+    }
+    const ext = product.previewImage.split(';')[0].match(/jpeg|png|jpg/)?.[0] ?? '';
+    const data = product.previewImage.replace(/^data:image\/\w+;base64,/, '');
+    const buf = Buffer.from(data, 'base64');
+    const storageFileName = `${PRODUCT_IMAGE_PREVIEW_PREFIX}${product.id}.${ext}`
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    fs.writeFile(`${this.imageStorageDirectory}/${storageFileName}`, buf, async (error) => {
```

Comment:
Are you writing the base64 into raw file here?

Reply:
Not exactly, it passes through the Buffer.from() first. I'm not sure if it has the same content after this binary -> string -> binary conversations, but it has exactly the same size as an original file.